MySQL database migration issue

### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

core

### Describe the bug

While deploying the update to 26.0.5 (updating from 25.0.6) an unexpected database migration issue was encountered.
This did not happen on my test environment, but it did happen with both production environments.

I believe this has to do with a small collation difference between those two environments: in test my database is `create schema keycloak collate utf8mb4_unicode_ci;`, where the production database use `utf8mb3_unicode_ci`.

However, only three tables are shown to use the mb4 collation and charset in my test environment:

- ORG
- ORG_DOMAIN
- REVOKED_TOKEN

Looking at https://www.keycloak.org/server/db#_configuring_unicode_support_for_a_mysql_database, utf8mb4 is not supported, which would mean that my test environment is running out of spec. But that was the only environment without issues.

Then I found this thread: https://groups.google.com/g/keycloak-user/c/9MhUfXwla3s?pli=1. Looks like `utf8mb4` should be used now with MySQL 8, and that the documentation needs to be updated?


### Version

26.0.5

### Regression

- [ ] The issue is a regression

### Expected behavior

The migration to be completed without issue.

### Actual behavior

```
keycloak-1  | 2024-11-15 13:43:56,257 ERROR [liquibase.changelog.ChangeSet] (main) ChangeSet META-INF/jpa-changelog-26.0.0.xml::26.0.0-org-group::keycloak encountered an exception.: liquibase.exception.DatabaseException: Illegal mix of collations (utf8mb3_general_ci,IMPLICIT) and (utf8mb3_unicode_ci,IMPLICIT) for operation '=' [Failed SQL: (1267) UPDATE keycloak.KEYCLOAK_GROUP SET TYPE = 1 WHERE NAME IN (SELECT ID FROM keycloak.ORG)]
```

```
keycloak-1  | 2024-11-15 13:44:11,230 ERROR [liquibase.changelog.ChangeSet] (main) ChangeSet META-INF/jpa-changelog-26.0.0.xml::26.0.0-org-group::keycloak encountered an exception.: liquibase.exception.DatabaseException: Duplicate column name 'TYPE' [Failed SQL: (1060) ALTER TABLE keycloak.KEYCLOAK_GROUP ADD TYPE INT DEFAULT 0 NULL]
```

### How to Reproduce?

I believe it should be reproducable when upgrading a 25 instance to 26 while using MySQL 8 where the complete database is using the `utf8mb3_unicode_ci` collation and `utf8mb3` character set.

### Anything else?

For now I fixed it by manually marking the migration as completed in the db and adding the `TYPE` column manually. As the specific table was empty, there was no need to run https://github.com/keycloak/keycloak/blob/cec081961b7116577dea764adc826b7005b23981/model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_0_0_OrganizationGroupType.java manually.

I created the database a long while ago using `CREATE DATABASE KEYCLOAK CHARACTER SET utf8 COLLATE utf8_unicode_ci;` and now after updating from 25 to 26 I get the error
```
keycloak-1  | 2024-11-27T13:03:13.227506589Z 2024-11-27 13:03:13,226 ERROR [liquibase.changelog.ChangeSet] (main) ChangeSet META-INF/jpa-changelog-26.0.0.xml::26.0.0-org-group::keycloak encountered an exception.: liquibase.exception.DatabaseException: Illegal mix of collations (utf8mb3_general_ci,IMPLICIT) and (utf8mb3_unicode_ci,IMPLICIT) for operation '=' [Failed SQL: (1267) UPDATE ee_keycloak.KEYCLOAK_GROUP SET TYPE = 1 WHERE NAME IN (SELECT ID FROM ee_keycloak.ORG)]
keycloak-1  | 2024-11-27T13:03:13.227534422Z 	at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:476)
keycloak-1  | 2024-11-27T13:03:13.227536506Z 	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:83)
keycloak-1  | 2024-11-27T13:03:13.227537506Z 	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:185)
keycloak-1  | 2024-11-27T13:03:13.227538714Z 	at liquibase.executor.AbstractExecutor.execute(AbstractExecutor.java:141)
keycloak-1  | 2024-11-27T13:03:13.227546297Z 	at liquibase.database.AbstractJdbcDatabase.executeStatements(AbstractJdbcDatabase.java:1189)
keycloak-1  | 2024-11-27T13:03:13.227547256Z 	at liquibase.changelog.ChangeSet.execute(ChangeSet.java:775)
keycloak-1  | 2024-11-27T13:03:13.227548089Z 	at liquibase.changelog.visitor.UpdateVisitor.executeAcceptedChange(UpdateVisitor.java:119)
keycloak-1  | 2024-11-27T13:03:13.227548964Z 	at liquibase.changelog.visitor.UpdateVisitor.visit(UpdateVisitor.java:68)
keycloak-1  | 2024-11-27T13:03:13.227549797Z 	at liquibase.changelog.ChangeLogIterator.lambda$run$0(ChangeLogIterator.java:131)
keycloak-1  | 2024-11-27T13:03:13.227550672Z 	at liquibase.Scope.lambda$child$0(Scope.java:191)
keycloak-1  | 2024-11-27T13:03:13.227551506Z 	at liquibase.Scope.child(Scope.java:200)
keycloak-1  | 2024-11-27T13:03:13.227552589Z 	at liquibase.Scope.child(Scope.java:190)
keycloak-1  | 2024-11-27T13:03:13.227553589Z 	at liquibase.Scope.child(Scope.java:169)
keycloak-1  | 2024-11-27T13:03:13.227554464Z 	at liquibase.changelog.ChangeLogIterator.lambda$run$1(ChangeLogIterator.java:120)
keycloak-1  | 2024-11-27T13:03:13.227555339Z 	at liquibase.Scope.lambda$child$0(Scope.java:191)
keycloak-1  | 2024-11-27T13:03:13.227556131Z 	at liquibase.Scope.child(Scope.java:200)
keycloak-1  | 2024-11-27T13:03:13.227557006Z 	at liquibase.Scope.child(Scope.java:190)
keycloak-1  | 2024-11-27T13:03:13.227557756Z 	at liquibase.Scope.child(Scope.java:169)
keycloak-1  | 2024-11-27T13:03:13.227558881Z 	at liquibase.Scope.child(Scope.java:257)
keycloak-1  | 2024-11-27T13:03:13.227559672Z 	at liquibase.Scope.child(Scope.java:261)
keycloak-1  | 2024-11-27T13:03:13.227560506Z 	at liquibase.changelog.ChangeLogIterator.run(ChangeLogIterator.java:89)
keycloak-1  | 2024-11-27T13:03:13.227561297Z 	at liquibase.command.core.AbstractUpdateCommandStep.lambda$run$0(AbstractUpdateCommandStep.java:113)
keycloak-1  | 2024-11-27T13:03:13.227562131Z 	at liquibase.Scope.lambda$child$0(Scope.java:191)
keycloak-1  | 2024-11-27T13:03:13.227562881Z 	at liquibase.Scope.child(Scope.java:200)
keycloak-1  | 2024-11-27T13:03:13.227563547Z 	at liquibase.Scope.child(Scope.java:190)
keycloak-1  | 2024-11-27T13:03:13.227564339Z 	at liquibase.Scope.child(Scope.java:169)
keycloak-1  | 2024-11-27T13:03:13.227565589Z 	at liquibase.command.core.AbstractUpdateCommandStep.run(AbstractUpdateCommandStep.java:111)
keycloak-1  | 2024-11-27T13:03:13.227566422Z 	at liquibase.command.core.UpdateCommandStep.run(UpdateCommandStep.java:105)
keycloak-1  | 2024-11-27T13:03:13.227567214Z 	at liquibase.command.CommandScope.execute(CommandScope.java:220)
keycloak-1  | 2024-11-27T13:03:13.227568089Z 	at liquibase.Liquibase.lambda$update$0(Liquibase.java:216)
keycloak-1  | 2024-11-27T13:03:13.227568839Z 	at liquibase.Scope.lambda$child$0(Scope.java:191)
keycloak-1  | 2024-11-27T13:03:13.227569547Z 	at liquibase.Scope.child(Scope.java:200)
keycloak-1  | 2024-11-27T13:03:13.227570256Z 	at liquibase.Scope.child(Scope.java:190)
keycloak-1  | 2024-11-27T13:03:13.227570964Z 	at liquibase.Scope.child(Scope.java:169)
keycloak-1  | 2024-11-27T13:03:13.227571672Z 	at liquibase.Liquibase.runInScope(Liquibase.java:1290)
keycloak-1  | 2024-11-27T13:03:13.227573172Z 	at liquibase.Liquibase.update(Liquibase.java:205)
keycloak-1  | 2024-11-27T13:03:13.227573922Z 	at liquibase.Liquibase.update(Liquibase.java:188)
keycloak-1  | 2024-11-27T13:03:13.227574631Z 	at liquibase.Liquibase.update(Liquibase.java:175)
keycloak-1  | 2024-11-27T13:03:13.227575381Z 	at org.keycloak.quarkus.runtime.storage.database.liquibase.QuarkusJpaUpdaterProvider.updateChangeSet(QuarkusJpaUpdaterProvider.java:191)
keycloak-1  | 2024-11-27T13:03:13.227576256Z 	at org.keycloak.quarkus.runtime.storage.database.liquibase.QuarkusJpaUpdaterProvider.update(QuarkusJpaUpdaterProvider.java:106)
keycloak-1  | 2024-11-27T13:03:13.227577131Z 	at org.keycloak.quarkus.runtime.storage.database.liquibase.QuarkusJpaUpdaterProvider.update(QuarkusJpaUpdaterProvider.java:84)
keycloak-1  | 2024-11-27T13:03:13.227578006Z 	at org.keycloak.quarkus.runtime.storage.database.jpa.QuarkusJpaConnectionProviderFactory.update(QuarkusJpaConnectionProviderFactory.java:281)
keycloak-1  | 2024-11-27T13:03:13.227578839Z 	at org.keycloak.quarkus.runtime.storage.database.jpa.QuarkusJpaConnectionProviderFactory.createOrUpdateSchema(QuarkusJpaConnectionProviderFactory.java:263)
keycloak-1  | 2024-11-27T13:03:13.227579756Z 	at org.keycloak.quarkus.runtime.storage.database.jpa.QuarkusJpaConnectionProviderFactory.postInit(QuarkusJpaConnectionProviderFactory.java:124)
keycloak-1  | 2024-11-27T13:03:13.227580631Z 	at org.keycloak.quarkus.runtime.integration.QuarkusKeycloakSessionFactory.init(QuarkusKeycloakSessionFactory.java:105)
keycloak-1  | 2024-11-27T13:03:13.227581422Z 	at org.keycloak.quarkus.runtime.integration.jaxrs.QuarkusKeycloakApplication.createSessionFactory(QuarkusKeycloakApplication.java:67)
keycloak-1  | 2024-11-27T13:03:13.227582297Z 	at org.keycloak.services.resources.KeycloakApplication.startup(KeycloakApplication.java:90)
keycloak-1  | 2024-11-27T13:03:13.227583047Z 	at org.keycloak.quarkus.runtime.integration.jaxrs.QuarkusKeycloakApplication.onStartupEvent(QuarkusKeycloakApplication.java:52)
keycloak-1  | 2024-11-27T13:03:13.227584047Z 	at org.keycloak.quarkus.runtime.integration.jaxrs.QuarkusKeycloakApplication_Observer_onStartupEvent_GNZ8m5QenZ9h9VNelo7awjUZFDE.notify(Unknown Source)
keycloak-1  | 2024-11-27T13:03:13.227584922Z 	at io.quarkus.arc.impl.EventImpl$Notifier.notifyObservers(EventImpl.java:351)
keycloak-1  | 2024-11-27T13:03:13.227585714Z 	at io.quarkus.arc.impl.EventImpl$Notifier.notify(EventImpl.java:333)
keycloak-1  | 2024-11-27T13:03:13.227586464Z 	at io.quarkus.arc.impl.EventImpl.fire(EventImpl.java:80)
keycloak-1  | 2024-11-27T13:03:13.227587214Z 	at io.quarkus.arc.runtime.ArcRecorder.fireLifecycleEvent(ArcRecorder.java:156)
keycloak-1  | 2024-11-27T13:03:13.227588006Z 	at io.quarkus.arc.runtime.ArcRecorder.handleLifecycleEvents(ArcRecorder.java:107)
keycloak-1  | 2024-11-27T13:03:13.227588881Z 	at io.quarkus.deployment.steps.LifecycleEventsBuildStep$startupEvent1144526294.deploy_0(Unknown Source)
keycloak-1  | 2024-11-27T13:03:13.227589756Z 	at io.quarkus.deployment.steps.LifecycleEventsBuildStep$startupEvent1144526294.deploy(Unknown Source)
keycloak-1  | 2024-11-27T13:03:13.227590589Z 	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
keycloak-1  | 2024-11-27T13:03:13.227591339Z 	at io.quarkus.runtime.Application.start(Application.java:101)
keycloak-1  | 2024-11-27T13:03:13.227592672Z 	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:119)
keycloak-1  | 2024-11-27T13:03:13.227593422Z 	at io.quarkus.runtime.Quarkus.run(Quarkus.java:71)
keycloak-1  | 2024-11-27T13:03:13.227594131Z 	at org.keycloak.quarkus.runtime.KeycloakMain.start(KeycloakMain.java:146)
keycloak-1  | 2024-11-27T13:03:13.227594881Z 	at org.keycloak.quarkus.runtime.cli.command.AbstractStartCommand.run(AbstractStartCommand.java:57)
keycloak-1  | 2024-11-27T13:03:13.227595631Z 	at picocli.CommandLine.executeUserObject(CommandLine.java:2030)
keycloak-1  | 2024-11-27T13:03:13.227596339Z 	at picocli.CommandLine.access$1500(CommandLine.java:148)
keycloak-1  | 2024-11-27T13:03:13.227597047Z 	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2465)
keycloak-1  | 2024-11-27T13:03:13.227597797Z 	at picocli.CommandLine$RunLast.handle(CommandLine.java:2457)
keycloak-1  | 2024-11-27T13:03:13.227598547Z 	at picocli.CommandLine$RunLast.handle(CommandLine.java:2419)
keycloak-1  | 2024-11-27T13:03:13.227599256Z 	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2277)
keycloak-1  | 2024-11-27T13:03:13.227599964Z 	at picocli.CommandLine$RunLast.execute(CommandLine.java:2421)
keycloak-1  | 2024-11-27T13:03:13.227600672Z 	at picocli.CommandLine.execute(CommandLine.java:2174)
keycloak-1  | 2024-11-27T13:03:13.227601381Z 	at org.keycloak.quarkus.runtime.cli.Picocli.run(Picocli.java:147)
keycloak-1  | 2024-11-27T13:03:13.227602089Z 	at org.keycloak.quarkus.runtime.cli.Picocli.parseAndRun(Picocli.java:135)
keycloak-1  | 2024-11-27T13:03:13.227602839Z 	at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:106)
keycloak-1  | 2024-11-27T13:03:13.227603547Z 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
keycloak-1  | 2024-11-27T13:03:13.227604339Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
keycloak-1  | 2024-11-27T13:03:13.227605214Z 	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:62)
keycloak-1  | 2024-11-27T13:03:13.227605964Z 	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:33)
keycloak-1  | 2024-11-27T13:03:13.227606714Z Caused by: java.sql.SQLException: Illegal mix of collations (utf8mb3_general_ci,IMPLICIT) and (utf8mb3_unicode_ci,IMPLICIT) for operation '='
keycloak-1  | 2024-11-27T13:03:13.227607547Z 	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:130)
keycloak-1  | 2024-11-27T13:03:13.227608256Z 	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
keycloak-1  | 2024-11-27T13:03:13.227609047Z 	at com.mysql.cj.jdbc.StatementImpl.executeInternal(StatementImpl.java:770)
keycloak-1  | 2024-11-27T13:03:13.227609756Z 	at com.mysql.cj.jdbc.StatementImpl.execute(StatementImpl.java:653)
keycloak-1  | 2024-11-27T13:03:13.227610464Z 	at io.agroal.pool.wrapper.StatementWrapper.execute(StatementWrapper.java:220)
keycloak-1  | 2024-11-27T13:03:13.227611214Z 	at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:470)
keycloak-1  | 2024-11-27T13:03:13.227612047Z 	... 76 more
```

--- a/model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_0_0_OrganizationGroupType.java
+++ b/model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_0_0_OrganizationGroupType.java
@@ -16,6 +16,7 @@
  */
 package org.keycloak.connections.jpa.updater.liquibase.custom;
 
+import liquibase.database.core.MySQLDatabase;
 import liquibase.exception.CustomChangeException;
 import liquibase.statement.core.RawSqlStatement;
 
@@ -26,7 +27,7 @@ public class JpaUpdate26_0_0_OrganizationGroupType extends CustomKeycloakTask {
         String groupTable = getTableName("KEYCLOAK_GROUP");
         String orgTable = getTableName("ORG");
 
-        if ("mariadb".equals(database.getShortName())) {
+        if (database instanceof MySQLDatabase) {
             statements.add(new RawSqlStatement("UPDATE " + groupTable + " SET TYPE = 1 WHERE CONVERT(NAME USING utf8) IN (SELECT CONVERT(ID USING utf8) FROM " +  orgTable + ")"));
         } else {
             statements.add(new RawSqlStatement("UPDATE " + groupTable + " SET TYPE = 1 WHERE NAME IN (SELECT ID FROM " +  orgTable + ")"));

